source/MaterialXGenOgsXml/GlslFragmentGenerator.cpp

```diff
@@ -237,31 +305,78 @@ ShaderPtr GlslFragmentGenerator::generate(const string& name, ElementPtr element
                 {
                     toVec3(outputSocket->getType(), finalOutput);
                 }
-                emitLine("return " + finalOutput, stage);
+                emitLine("return " + finalOutput, pixelStage);
             }
         }
         else
         {
-            string outputValue = outputSocket->getValue() ? _syntax->getValue(outputSocket->getType(), *outputSocket->getValue()) : _syntax->getDefaultValue(outputSocket->getType());
+            const string outputValue = outputSocket->getValue()
+                ? _syntax->getValue(outputSocket->getType(), *outputSocket->getValue())
+                : _syntax->getDefaultValue(outputSocket->getType());
+
             if (!outputSocket->getType()->isFloat3())
             {
                 string finalOutput = outputSocket->getVariable() + "_tmp";
-                emitLine(_syntax->getTypeName(outputSocket->getType()) + " " + finalOutput + " = " + outputValue, stage);
+                emitLine(_syntax->getTypeName(outputSocket->getType()) + " " + finalOutput + " = " + outputValue, pixelStage);
                 toVec3(outputSocket->getType(), finalOutput);
-                emitLine("return " + finalOutput, stage);
+                emitLine("return " + finalOutput, pixelStage);
             }
             else
             {
-                emitLine("return " + outputValue, stage);
+                emitLine("return " + outputValue, pixelStage);
             }
         }
     }
 
     // End function
-    emitScopeEnd(stage);
+    emitScopeEnd(pixelStage);
 
     // Replace all tokens with real identifier names
-    replaceTokens(_tokenSubstitutions, stage);
+    replaceTokens(_tokenSubstitutions, pixelStage);
+
+    // Now emit private uniform definitions to a special stage which is only
```

Comment:
Minor: Should this only be called if MaterialXCross build flag is on?

Reply:
Right, if cross-compilation is disabled then generating this stage is redundant. However the current motivation for the macro is that we can't build `MaterialXCross` in CI and this code builds fine without `MaterialXCross`, so leaving it enabled helps cover it by CI. I think ideally we should just avoid cross-compilation at runtime if VP2 is in an OpenGL mode because the mode can only be switched by restarting Maya. Let me make this improvement in a future PR.

Reply:
Sounds good.